### PR TITLE
#166840916 Implement user email verification notification on successful signup

### PR DESCRIPTION
**Pull request by timi-codes**
#### What does this PR do?
- add the feature for user to receive a verification link via email upon successful registration
#### Description of Task to be completed?
  -  SendGrid account
  -  Setup email template on SendGrid
  -  Implement mailer function to send mail

#### How should this be manually tested?
 - Clone the project, run `npm install` to install all dependencies and  run `npm start`
 - When a user signs up for an account using `/api/v1/auth/signup` endpoints they get an email address telling them to verify their account through the click of a button.
#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#166840916](https://www.pivotaltracker.com/story/show/166840916)

### CODE REVIEW

File: src/helpers/mailer.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { config } from 'dotenv';
+import sgMail from '@sendgrid/mail';
+
```
Comment (kechyy) [inline src/helpers/mailer.js:3]: @timi-codes I think Lawrence is also implementing email sending and you may want to rub mind with him on this. I feel we could have a generic method that does that so anyone implementing email sending could use that .

**Review by kechyy — COMMENTED** [review]
@timi-codes  I like the concept of the sendGrid template to manage both HTML and plain text templates inside the SG dashboard rather than in your code .
But I feel you need to rub mind with @olorunwalawrence to agree on the one to use since you are both implementing email sending task.

File: src/helpers/mailer.js (hunk 1 above)
Comment (timi-codes) [reply src/helpers/mailer.js:3]: Thank you @kechyy . We will work on this

File: src/test/user.spec.js [hunk 2]
```diff
@@ -37,7 +42,18 @@ describe('Test user login, signup and account verification', () => {
...
+
+      nock('https://api.sendgrid.com')
+        .post('/api/mail.send.json')
+        .reply(200, {
+          message: 'success'
+        });
+    });
+
+    it('should throw an error when user with an unverified account wants to signup', done => {
       const newUser = {
```
Comment (olorunwalawrence) [inline src/test/user.spec.js]: I don't know the reason there are two it statement in these lines of code please could you visit this?

File: src/test/user.spec.js [hunk 3]
```diff
@@ -37,7 +42,18 @@ describe('Test user login, signup and account verification', () => {
         });
     });
 
-    it('it should throw an error when user with an unverified account wants to signup', done => {
+    it('should mock a sendgrid api request', async () => {
```
Comment (olorunwalawrence) [inline src/test/user.spec.js]: I think this will be sending an email anytime the test runs will this be ok?

**Review by tohbay — COMMENTED** [review]
Good work here 👍. I noticed you removed the details in this file as it will pose a security risk having them there.

File: src/controllers/user.controller.js [hunk 4]
```diff
@@ -54,7 +54,16 @@ class UserController {
...
+      const { protocol } = req;
+      mailer({
+        name: username,
+        receiver: email,
+        subject: 'Welcome to ErrorSwag',
+        templateName: 'confirm_account',
+        confirm_account_url: `${protocol}://${req.get(
+          'host'
+        )}/api/v1/auth/verify/${token}`
+      });
```
Comment (tohbay) [inline src/controllers/user.controller.js:67]: Good work here 👍. However, as a `User`, when I `cloned` the `repo` and `signed up` with my `Gmail` account, I received no email but I got a `response body` from `Postman` as shown below. 
image

Is there something I didn't do that made me not get the email sent upon registration?

Comment (timi-codes) [reply src/controllers/user.controller.js:67]: You may need to check your spam if it's not sent to your inbox

File: src/controllers/user.controller.js [hunk 5]
```diff
@@ -96,7 +105,18 @@ class UserController {
...
+            mailer({
+              name: username,
+              receiver: email,
+              subject: 'Welcome to ErrorSwag',
+              templateName: 'confirm_account',
+              confirm_account_url: `${protocol}://${req.get(
+                'host'
+              )}/api/v1/auth/verify/${token}`
+            });
+
```
Comment (tohbay) [inline src/controllers/user.controller.js:119]: Nice here 👍, however as I have pointed out earlier on the signup section. The image below shows the response I got upon login:
image

As a user, I am expected to get an email notification but I didn't get. Is there something I am expected to do so I can get the email notification?

File: src/helpers/mailer.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      name,
+      confirm_account_url: data.confirm_account_url,
+      reset_password_url: data.reset_password_url
+    }
+  };
+
+  sgMail.send(msg);
+};
+
+export default mailer;
```
Comment (tohbay) [inline src/helpers/mailer.js:30]: Good job here 👍. going forward, you may want to take another look at this to know if it is responsible for not sending the notification email; or is there something I am missing on my own part?

File: src/test/user.spec.js [hunk 7]
```diff
@@ -37,7 +42,18 @@ describe('Test user login, signup and account verification', () => {
...
+      const secret = process.env.SENDGRID_API_KEY;
+      sgMail.setApiKey(secret);
+
+      nock('https://api.sendgrid.com')
+        .post('/api/mail.send.json')
+        .reply(200, {
+          message: 'success'
+        });
+    });
+
```
Comment (tohbay) [inline src/test/user.spec.js:73]: I noticed you just mocked the sendgrid API but there is no assertion to test the expected response for an email notification. You may have to look at [this](https://github.com/sendgrid/sendgrid-nodejs/issues/141#issuecomment-63570373) for some insight.

File: src/test/user.spec.js [hunk 8]
```diff
@@ -17,7 +22,7 @@ describe('Test user login, signup and account verification', () => {
   let validToken;
 
   describe('POST /auth/signup', () => {
-    it('it should create a user account', done => {
+    it('should create a user account', done => {
```
Comment (tohbay) [inline src/test/user.spec.js]: Nice test set up 👍for a test-driven development (TDD) approach, however, there is less confidence in this test as there are no assertions to test for the properties (behaviour) of a user (username, email, role, status) as shown in the response body.

This applies to the rest of the tests. 
Going forward, there is room for improvement on these tests in order to give confidence in the feature implemented.

File: .env.example [hunk 9]
```diff
@@ -1,14 +1,14 @@
-NODE_ENV=development
-DB_USERNAME=timitejumola
-DB_PASSWORD=Oluwatimileyin9595
-DB_DATABASE=errorswag
-DB_HOST=localhost
+NODE_ENV=
```
Comment (tohbay) [inline .env.example:1]: Good job 👍  @timi-codes,
I noticed you removed the details contained in the `.env.example` file to maintain security in the application.

However, the PR description is not good enough to guide a user on how to test the feature implemented. Hence there is room for improvement. 
Going forward, you can look at the link [here](https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices) for some insights.

There are some reviews above.

**Review by codexempire — COMMENTED** [review]
Nice Job @timi-codes, Please work on the reviews given here

**Review by kechyy — APPROVED** [review]
Good Job @timi-codes . 
Thanks for acting on the feedback . 
It looks good to me.

**Review by codexempire — APPROVED** [review]
Well done

**Review by tolumide-ng — APPROVED** [review]
Good job @timi-codes

File: src/test/user.spec.js [hunk 10]
```diff
@@ -37,7 +42,35 @@ describe('Test user login, signup and account verification', () => {
...
+        });
+
+      const res = await chai
+        .request('https://api.sendgrid.com')
+        .post('/api/mail.send.json');
+      expect(res.body.message).to.include('success');
+      expect(res.status).to.eql(200);
+    });
+
+    it('should mock a failed sendgrid api request', async () => {
```
Comment (segunolalive) [inline src/test/user.spec.js]: 😅 
The test block should say what is done by the `API` being tested. Your `API` doesn't **mock** sendgrid. You api sends requests to sendgrid and **handles** both success and failure cases. That's what your test should say.

The assertions should demonstrate this as well.

File: src/test/user.spec.js [hunk 11]
```diff
@@ -37,7 +42,35 @@ describe('Test user login, signup and account verification', () => {
...
+      nock('https://api.sendgrid.com')
+        .post('/api/mail.send.json')
+        .reply(400, {
+          message: 'failed'
+        });
+
+      const res = await chai
+        .request('https://api.sendgrid.com')
+        .post('/api/mail.send.json');
+      expect(res.body.message).to.include('failed');
```
Comment (segunolalive) [inline src/test/user.spec.js]: I'd prefer that you assert the actual error message not just that it includes a  word or phrase.

File: src/test/user.spec.js (hunk 10 above)
Comment (segunolalive) [reply src/test/user.spec.js]: You may look at this for [ideas](https://github.com/segunolalive/helloBooks/blob/development/server/test/controllers/userController.spec.js#L191)

**Review by segunolalive — CHANGES_REQUESTED** [review]

File: src/test/user.spec.js [hunk 12]
```diff
@@ -38,66 +42,88 @@ describe('Test user login, signup and account verification', () => {
...
-        .send(user)
-        .end((err, res) => {
-          expect(res.status).to.equal(400);
-          expect(res.body).to.have.property('status');
-          expect(res.body.error).to.equal('email must be a valid email');
-          done();
-        });
+        .send(user);
+      expect(res.status).to.equal(400);
+      expect(res.body).to.have.property('status');
```
Comment (segunolalive) [inline src/test/user.spec.js]: remove property status

File: src/test/user.spec.js [hunk 13]
```diff
@@ -109,14 +135,14 @@ describe('Test user login, signup and account verification', () => {
...
-          expect(res.body.error).to.equal(
+          expect(res.body.message).to.equal(
             'password length must be at least 8 characters long'
           );
           done();
         });
     });
 
-    it('it should throw a wrong credential is passed', done => {
+    it('should throw a wrong credential is passed', done => {
```
Comment (segunolalive) [inline src/test/user.spec.js]: `should throw a`  change to -> `should throw if`

File: src/test/user.spec.js [hunk 14]
```diff
@@ -133,104 +159,126 @@ describe('Test user login, signup and account verification', () => {
...
+        .send({
+          email: user.email,
+          password: user.password
         });
+
+      expect(res.status).to.equal(200);
+      expect(res.body).to.have.property('status');
+      expect(res.body.data)
+        .to.have.property('status')
+        .to.equal('active');
```
Comment (segunolalive) [inline src/test/user.spec.js]: show that a user is logged in by testing that a token was returned.

You can assert the validity of the token, the contents of the token etc. This is the login contract that the app exposes. This is how we know that a user is logged in.

File: src/test/user.spec.js [hunk 15]
```diff
@@ -239,7 +287,6 @@ describe('GET api/v1/users', () => {
         expect(res.status).to.equal(200);
         expect(res.body.data).to.be.an('array');
         expect(res.body.data).to.have.lengthOf(4);
```
Comment (segunolalive) [inline src/test/user.spec.js]: Assert that the items gotten are actually user objects e.g check that they have the required user fields.

**Review by segunolalive — CHANGES_REQUESTED** [review]

**Pull request merged**